tools/internal/parser/validate.go

```diff
@@ -37,3 +45,137 @@ func (p *parser) requirePrivateDomainEmailContact() {
 		}
 	}
 }
+
+const (
+	amazonSuperblockStart = "Amazon : https://www.amazon.com"
+	amazonSuperblockEnd   = "concludes Amazon"
+)
+
+// requireSortedPrivateSection verifies that the blocks in the private
+// domains section is sorted according to PSL policy.
+func (p *parser) requireSortedPrivateSection() {
+	type chunk struct {
+		Name     string
+		Suffixes []Suffixes
+	}
+
+	coll := newCollator()
+	cmp := func(a, b chunk) int { return coll.Compare(a.Name, b.Name) }
+
+	// Amazon has a semi-automated "superblock" of suffix blocks,
+	// which are in the PSL at the correct sort location for "Amazon",
+	// but are not correctly interleaved with other non-Amazon
+	// blocks. This causes a lot of churn in the sorting for what is
+	// currently an edge case.
+	//
+	// For a smooth transition, this validation groups all the Amazon
+	// blocks into a single chunk, and sorts that entire chunk
+	// relative to the rest of the file. Ordering within the Amazon
+	// superblock is not currently enforced, only its order relative
+	// to other blocks.
+	//
+	// TODO: rework the parse tree structure so these "superblocks"
+	// can be tracked more cleanly than this.
+	var chunks []chunk
+	inAmazonBlock := false
+	for _, block := range allBlocksInPrivateSection(&p.File) {
+		if comm, ok := block.(Comment); ok {
+			if !inAmazonBlock && strings.Contains(comm.Text(), amazonSuperblockStart) {
+				inAmazonBlock = true
+				chunks = append(chunks, chunk{
+					Name: "Amazon",
+				})
+			} else if inAmazonBlock && strings.Contains(comm.Text(), amazonSuperblockEnd) {
+				inAmazonBlock = false
+			}
+			continue
+		}
+
+		suffixes, ok := block.(Suffixes)
+		if !ok {
+			continue
+		}
+
+		if inAmazonBlock {
+			last := len(chunks) - 1
+			chunks[last].Suffixes = append(chunks[last].Suffixes, suffixes)
+			continue
+		}
+
+		chunks = append(chunks, chunk{
+			Name:     suffixes.Entity,
+			Suffixes: []Suffixes{suffixes},
+		})
+	}
+
+	// Calculate the longest non-decreasing subsequence of
+	// chunks. This gives us the portion of the list that is already
+	// in the correct order. From there, we can look at the rest of
+	// the elements and report sort errors.
+	sorted := slice.LNDSFunc(chunks, cmp)
+	insertIdx := func(chunk chunk) int {
+		ret, _ := slices.BinarySearchFunc(sorted, chunk, cmp)
+		return ret
+	}
+
+	// Scan through chunks and report errors for misplaced items.
+	byTargetIdx := map[int][]chunk{}
```

Comment:
Maybe there are some comments that we could add to make this more understandable? I assume the high-level idea is to check for each chunk if it's part of the longest sorted sequence and otherwise report where it ought to be in that sequence but the index math is really hard to follow.

Reply:
I ended up redoing this entire piece of logic to report errors that are easier to act upon, because my initial attempt was actually hard for users to execute. It will become irrelevant soon once I implement pslfmt, but I've added verbose explanations of how the logic works.